KEP-4818: Allow zero value for Sleep Action of PreStop Hook

sreeram-venkitesh opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Added the kep.yaml files and the initial readme file for KEP 4818.

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4818

<!-- other comments or additional information -->
- Other comments: [Discussion doc](https://docs.google.com/document/d/1o01gH2ELPY-kjwgemjg3BbWiykx37ag-4VYwwJ1jBTw/edit?usp=sharing)

kannon92 commented:
/cc

kannon92 commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,963 @@
+
+## Release Signoff Checklist
+
+<!--
```
As you fill at sections, I recommend removing the comments. It helps keep the size of the diff small and makes it easier to review the markdown.

kannon92 reviewed (commented):
You will need a PRR file also.

kannon92 commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,564 @@
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
```
Please fill out these sections.

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,564 @@
+
+The Sleep action is implemented with the time package from Go’s standard library. The `time.After()` which is used to implement the sleep permits a zero sleep duration. A negative or a zero sleep duration will cause the function to return immediately and function like a no-op.
+
+The implementation in KEP 3960 supports only non-zero values for the sleep duration. It is semantically correct to support a zero value for this field since time.After() also supports zero and negative durations. Negative values as well as zero have the same effect with time.After(), they both return immediately. We don’t need to support negative values since they have the same effect as setting  ...
```
can you please expand on whether kubelet accepts `0` today and it is only API server validaiton or kubelet also validates the value?

sreeram-venkitesh replied:
Currently, kubelet accepts `0`. No validation is done at the kubelet level. The feature gate for `PodLifecycleSleepAction` is used inside `runSleepHandler` function and it returns nil if the feature is disabled. We can do a similar check at the kubelet level but it doesn't really make sense since a sleep of zero seconds would be similar to returning `nil` right? Since the logic in kubelet already supports zero value (`time.After(0)`), I feel that we don't need the feature gate in kubelet. Also the feature gate check for `PodLifecycleSleepAction` would be removed in v1.32 since its planning to graduate to stable. Please let me know what you think!

[Relevant function in kubelet's lifecyle/handlers.go](https://github.com/AxeZhan/kubernetes/blob/3a96afdfefdf329c637623ae31a61d20dbdb0393/pkg/kubelet/lifecycle/handlers.go#L129-L141)

```go
func (hr *handlerRunner) runSleepHandler(ctx context.Context, seconds int64) error {
	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepAction) {
		return nil
	}
	c := time.After(time.Duration(seconds) * time.Second)
	select {
	case <-ctx.Done():
		// unexpected termination
		return fmt.Errorf("container terminated before sleep hook finished")
	case <-c:
		return nil
	}
}
```

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,564 @@
+
+### Non-Goals
+
+N/A
```
support `Infinity` should not be a goal

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,564 @@
+- When the feature gate is disabled:
+  - (a) New resources will no longer allow setting zero as the sleep duration second for the PreStop hook. (no change to current validation)
+  - (b) Existing resources cannot be updated to have a sleep duration of zero seconds
+  - (c) Existing resources with a PreStop sleep duration set to zero will continue to run and use a sleep duration of zero seconds. These resources however cannot be updated without also updating the sleep duration to a non-zero value, since the validation would now fail.
```
> These resources however cannot be updated

So status updates will not be allowed? You may want to be more precise on what updates will not be allowed

sreeram-venkitesh replied:
Status updates are allowed. [Ref](https://docs.google.com/document/d/1o01gH2ELPY-kjwgemjg3BbWiykx37ag-4VYwwJ1jBTw/edit?disco=AAABWDp70to). Also [discussion in the PR](https://github.com/kubernetes/kubernetes/pull/127094#discussion_r1764005018). I need to update the KEP with this. Will do it today.

SergeyKanzhelev replied:
yep, as expected. I just pointing out that this statement is quite scary and should be clarified.

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/kep.yaml:
```diff
@@ -0,0 +1,25 @@
+id: ""
```
please fill this file up.

sreeram-venkitesh commented:
Addressed all the above comments, please let me know if there are any more tests that I can add!

SergeyKanzhelev commented on keps/prod-readiness/sig-node/4818.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 4818
+alpha:
+  approver: ""
```
pick any. Ideally reuse one from the previous KEP for sleep action

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,580 @@
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PodLifecycleSleepActionAllowZero
+  - Components depending on the feature gate:
```
api-server?

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/kep.yaml:
```diff
@@ -0,0 +1,31 @@
+creation-date: "2024-09-16"
+
+reviewers: []
+approvers: []
```
you can put me

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/kep.yaml:
```diff
@@ -0,0 +1,31 @@
+status: implementable
+creation-date: "2024-09-16"
+
+reviewers: []
```
maybe @kannon92 ?

kannon92 replied:
happy too.

SergeyKanzhelev reviewed (commented):
lgtm, with couple of items to address.

SergeyKanzhelev commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,580 @@
+<!--
```
since this is just a small additon to the previos KEP, I think it will be OK to add the OWNERS file: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3960-pod-lifecycle-sleep-action/OWNERS

sreeram-venkitesh replied:
I've added the file. Can I know more about why this file is needed at a KEP level? I haven't seen it much before.

sreeram-venkitesh commented:
@SergeyKanzhelev @kannon92 Thank you so much! I've resolved the comments and added you two as approver and reviewer. I still need to add the PRR approver, but I first pinged wojtek-t (who is the PRR approver for KEP 3960) in Slack to confirm with him first that I can add him as the PRR approver for this KEP too. I will update this field once he confirms.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

You still need PRR approver to be listed in 4818.yaml

SergeyKanzhelev commented:
/lgtm

SergeyKanzhelev commented:
/assign @wojtek-t

kannon92 commented:
@SergeyKanzhelev I don't think your owner file worked here.

@mrunalp @dchen1107 can you help here?

SergeyKanzhelev commented:
> @SergeyKanzhelev I don't think your owner file worked here.
> 
> @mrunalp @dchen1107 can you help here?

Yes, it will start working once merged. Also we need @wojtek-t 's PRR review

wojtek-t commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,580 @@
+
+If the kube-apiserver is downgraded to a version where the feature gate is not supported (<v1.32), no new resources can be created with a PreStop sleep duration of zero seconds. Existing resources will continue to function, but they cannot be updated since the validation for the sleep duration would fail if the value is set to zero. Cluster admins can update their resources to have a non-zero val ...
+
+If the feature gate is turned off after being enabled, no new resources can be created with PreStop sleep duration of zero seconds. Existing resources will continue to function, but as discussed in the Proposal section, the sleep duration will need to be updated to a non-zero value if the resource has to be updated.
```
This is outdated vs proposal section - existing resources should continue to work as described in the proposal (in particular this one: "Existing resources with a PreStop sleep duration set to zero will continue to run and use a sleep duration of zero seconds. These can be updated and the zero sleep duration would continue to work.")

Plese update

wojtek-t commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,580 @@
+
+###### Are there any tests for feature enablement/disablement?
+
+For the parent KEP, unit tests for the `switch` of the feature gate were added in `pkg/registry/core/pod/strategy_test`. We can add similar tests for the new feature gate as well.
```
Please do that for Alpha.

wojtek-t reviewed (commented):
One minor comment - other than that LGTM from PRR perspective.

pacoxu commented on keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md:
```diff
@@ -0,0 +1,580 @@
++		}
++	} else {
++		if gracePeriod != nil && sleep.Seconds <= 0 || sleep.Seconds > *gracePeriod {
++			invalidStr := fmt.Sprintf("must be greater than 0 and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
```
Should we include a suggestion to include the new FG that can allow zero in the error message?

sreeram-venkitesh commented:
@wojtek-t: I've fixed the section on downgrading, I missed that earlier! Thanks for catching it. For https://github.com/kubernetes/enhancements/pull/4855#discussion_r1778207898, I've added a line mentioning this test case under the list of unit tests planned to be implemented in alpha.

@pacoxu: I've updated the validation error message to mention the feature gate. Will update this in https://github.com/kubernetes/kubernetes/pull/127094.

Please take a look! Thanks folks!

SergeyKanzhelev reviewed (approved):
/lgtm

pacoxu commented:
/lgtm

wojtek-t commented:
/lgtm
/approve PRR
